design/aws-ecs-priv-registry-support.md

```diff
@@ -0,0 +1,146 @@
+
+Therefore, to support the use of private registries we would need to add credentials to the `DockerHubImage` type (as optional), and add a new `PrivateImage` type which could house credentials (required) and the private registry URI.
+
+We should also have a unified way to define repository credentials which can be passed to the constructors of both image types.  
+
+## Code changes
+
+Given the above, we should make the following additions to support private registry authentication:
+
+1. Define `RepositoryCredentials` interface & class, add to `IContainerImage` 
+2. Update `DockerHubImage` construct to optionally accept and set `RepositoryCreds` 
+3. Add new `PrivateImage` class which implements `IContainerImage` and requires `RepositoryCreds`
+
+
+# Part 1: How to define registry credentials 
+
+For extensibility, we can define a new `IRepositoryCreds` interface with a "credentialsParamater" string and a new `RepositoryCreds` class which satisfies it using specific constructs (e.g., "fromSecret").
+
+```ts
+export interface IRepositoryCreds {
+    readonly credentialsParameter: string
+}
+
+export class RepositoryCreds {
+    public readonly credentialsParameter: string;
+    private readonly secret: secretsManager.Secret;
+
+    public static fromSecret(secret: secretsManager.Secret) {
+      // sets credentialsParameter from secret.secretArn 
+    }
+
+    public static withCredentialsParameter(param: string) {
+      // sets credentialsParameter as provided string
+    }
+
+    public bind(containerDefinition: ContainerDefinition): void {
+      if (this.secret !== null) {
+        this.secret.grantRead(containerDefinition.taskDefinition.obtainExecutionRole());
+      }
+  }
+}
+```
+
+The `IContainerImage` interface will be updated to include an optional `repositoryCredentials` property:
+```ts
+export interface IContainerImage {
+  // ...
+  readonly imageName: string;
+
+  /**
+   * NEW: The credentials required to access the image
+   */
+  readonly repositoryCredentials?: IRepositoryCreds;
+
+  // ...
+  bind(containerDefinition: ContainerDefinition): void;
+}
+```
+
+
+## Part 2: Extend `DockerHubImage` class to include optional creds
+
+The `DockerHubImage` construct, which currently takes a string for image name, can be extended to take in an (optional) "repositoryCredentials" field of type `IRepositoryCreds`:
+```ts
+export class DockerHubImage implements IContainerImage {
+  // add repositoryCredentials to constructor
+  constructor(public readonly imageName: string, public readonly repositoryCredentials: RepositoryCreds) {
+  }
+
+  public bind(_containerDefinition: ContainerDefinition): void {
+    // Nothing to do
```

Comment:
`Nothing to do` is no longer true, right? We should be calling `this.repositoryCredentials.bind()` in here if I'm understanding correctly?

Reply:
You're right; I have `bind()` defined on `RepositoryCreds` but I suppose it needs to be actually called at some point :P